### ISSUE TITLE
Feature Request: terraform support

### ISSUE THREAD

**Issue by fugkco** [updatecli/updatecli#534]
### Is your feature request related to a problem?

Terraform modules can get far behind.

### Solution you'd like

I'd like to use updatecli to update two different things for a Terraform workspace.

### Terraform Providers
The first is Terraform providers, which provide backends for Terraform code. These are usually APIs such as AWS, Azure, or others such as Kubernetes, and many more. A full list is on the [Terraform registry](https://registry.terraform.io/browse/providers). As far as the source code that needs to be updated, I think we can simply target terraform 0.13 and up which allows you to specify providers as below:

```hcl
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 3.0"
    }
  }
}
```

With this, we can target the version and see what the latest "source" provider is using the registry API. E.g. for `hashicorp/aws` provider, we can query the search:
```
$ curl 'https://registry.terraform.io/v2/providers?filter%5Bnamespace%5D=hashicorp&filter%5Bname%5D=aws&filter%5Bunlisted%5D=true&filter%5Bwithout-versions%5D=true' | jq
{
  "data": [
    {
      "attributes": {
...
      },
      "id": "323",
      "links": {
        "self": "/v2/providers/323"
      },
      "type": "providers"
    }
  ],
  "links": {
...
  },
  "meta": {
...
  }
}

```
This will give us an ID, which we can then use to query all the versions:
```
curl 'https://registry.terraform.io/v2/providers/323?include=provider-versions' | jq
```

We can use [Hashicorp's HCL](https://github.com/hashicorp/hcl) package to parse and write the new version. This is the package that is used by Terraform itself.

### Terraform Modules
On the other side, there is also Terraform modules. Modules are essentially bundles of Terraform code (resources, data sources etc.) that are combined into a single module code. An example of a module call is:

```hcl
module "vpc" {
  source  = "terraform-aws-modules/vpc/aws"
  version = "3.11.5"
  # other attributes for the aws vpc module.
}
```

They here is retrieving the `version` attribute (or the `source` attribute in certain cases), and subsequently retrieving the latest version accordingly. Getting the latest version of a module is a bit more nuanced than the provider versions, because modules can be retrieved from different places. Git repos, Terraform registry, Terraform Enterprise private registry, Terraform Cloud private registry, [Terraform's public registry](https://registry.terraform.io/browse/modules) and likely other places I'm not aware of.

Terraform internally uses Hashicorp's [go-getter](https://github.com/hashicorp/go-getter) package, which allows for parsing the the `source` attribute as above. We can utilise it to figure out where to get the latest version from, and subsequently retrieve the relevant latest version, and compare. The similar to the providers, we can the HCL package to the write it back to file.

**Comment by olblak** [comment]
Hi Fugko,

Thank you for taking the time to write this down.
I think I would firstly focus on the provider, and then the module 
I would split the task in the following order: 

1. Write the code that retrieves the version from the Terraform API
2. Implement the resource interface, especially with the source function so updatecli knows how to retrieve the latest version.
3. Implement the HCL resource <- Temporarily we can rely on the file resource as a fallback.

I think I would create two new Golang packages
1. pkg/plugins/resources/terraform/module
2. pkg/plugins/resources/terraform/provider

A "resource" contains the following files
* `condition.go`: That implements the updatecli resource interface with the condition function, with a placeholder if needed yet
* `condition_test.go`: That contains the unit test related to `condition.go` 
* `source.go`:  That contains the code that implements the resource interface with the source function
* `source_test.go`: That contains the unit tests related to the `source.go`
* `target.go`:  that contains the code that implements the resource interface with the target function.
* `target_test.go`: that contains the unit tests related to the `target.go`
* `main.go`: that contains the code specific to the package.

The `main.go` also contains the `type Spec struct` which is used to parse an Updatecli configuration

You can have an example of a resource by looking at the maven one, on https://github.com/updatecli/updatecli/tree/main/pkg/plugins/resources/maven

By using unit test, you can test the provider resource without having to build the updatecli.

Once the package is ready, you can tell updatecli how to map "terraform/provider" to it by updating this file 
https://github.com/updatecli/updatecli/blob/main/pkg/core/pipeline/resource/main.go
 
Supporting HCL syntax is being discussed in this issue https://github.com/updatecli/updatecli/issues/234

**Comment by dduportal** [comment]
> because of #1424, there is also https://github.com/minamijoyo/tfupdate
> 
> Since this does the right thing (easy to test because terraform released 5.10.0 today (2023-07-31)
> 
> versions.tf says
> 
> ```
> terraform {
>   required_providers {
>     aws = {
>       source  = "hashicorp/aws"
>       version = "5.9.0"
>     }
>   }
> }
> ```
> 
> Which means that this does the "right thing" (for my simplistic view of the right thing) when trying to envisage how updatecli might update the providers.
> 
> ```
> bsh ❯ tfupdate release latest terraform-providers/terraform-provider-aws
> 5.10.0
> bsh ❯ tfupdate provider aws -v "5.10.0" versions.tf
> bsh ❯ tfupdate lock --platform=windows_amd64 --platform=linux_amd64 --platform=darwin_amd64 --platform=darwin_arm64 .
> ```
> 
> If you have versions.tf with `~> 5.0` then the `tfupdate lock` doesn't appear to update the hashes.
> 
> git diff now shows
> 
> ```
>  provider "registry.terraform.io/hashicorp/aws" {
> -  version     = "5.9.0"
> -  constraints = "~> 5.0"
> +  version     = "5.10.0"
> +  constraints = "5.10.0"
>    hashes = [
> -    "h1:TjctPnxYpg1RZaU1dGW8BSvdmf0t0jsNGU1rEmNIXns=",
> - ... (boring)
> +    "h1:AgF54/79Nb/oQjbAMMewENSIa1PEScMn20Xa91hZR2g=",
> +    "h1:LxQU2EauL2DKoWdnFtVZEt3lAWtN8neMWCj72/wdbBc=",
> +    "h1:g0NYapqztQLdHycec4tlBGNQJTUXUY0Xy5/XUgjAD7U=",
> +    "h1:ll2mC5mMF+Tm/+tmDQ6p6h3oAFpMSbZsA54STMZegwI=",
> ```
> 
> Would this help with your provider use-case @dduportal (@mcwarman because I'm sure you're itching to raise another PR)

That looks really good yes, thanks for the explanation 👍 

One point to have in mind: there are 2 schools when it is about keeping versions and we should cover both cases

- The "define versions in the tf files and keep the lock file updated"
- Only uses the lock file (this is what I tend to do for info)

Does it make sense with the `tfupdate` case?

**Comment by quotidian-ennui** [comment]
@dduportal 

- define versions in tf files and keep the lock file update is definitely catered for right now, and would be a simple wrapper a-la #1424 

If you just have a .terraform.lock.hcl file, then the tfupdate executable doesn't do anything if you update the versions + contraints. (i.e. `"h1:Tjc"` is still the first hash); so the executable variant requires a providers block?

But w/o looking at the code, and just going on the behaviour of the executable is potentially the wrong way to think about it

For your mode that only has a lock file, then we would have to

- use the hcledit go-mod to update `versions` && `constraints` with the path "provider/registry.terraform.io/hashicorp/aws" (after all it's still a valid HCL config file).
- use the tfupdate go-mod to compute the hashes and update them

(that's a more involved than a simple wrapper but potentially doable)

**Comment by mcwarman** [comment]
With the new `hcl` resource plugin (#1424) technically updating a terraform module from the registry and github is now possible

### Terraform Modules - Registry

```yaml
name: Terraform AWS Modules

sources:
  vpc:
    name: Get version from registry
    kind: json
    spec:
      file: https://registry.terraform.io/v1/modules/terraform-aws-modules/vpc/aws
      query: versions.[*]
      versionfilter:
        kind: semver

targets:
  local:
    name: Set VPC module version
    kind: hcl
    sourceid: vpc
    spec:
      file: vpc.tf
      path: module.vpc.version
```

This could be wrapped for a better user experience.

### Terraform Modules - GitHub

```yaml
name: Terraform AWS Modules

sources:
  vpc:
    kind: githubrelease
    spec:
      owner: terraform-aws-modules
      repository: terraform-aws-vpc
      token: "{{ requiredEnv `GITHUB_TOKEN` }}"
      versionfilter:
        kind: latest

targets:
  local:
    name: Set VPC module version
    kind: hcl
    sourceid: vpc
    transformers:
      - addprefix: "git::https://github.com/terraform-aws-modules/terraform-aws-vpc.git?ref="
    spec:
      file: vpc.tf
      path: module.vpc.source
```

Both potential candidates for `examples` folder.

**Comment by mcwarman** [comment]
@olblak and @dduportal to consider this complete how about:

- [x] HCL Updates #1424 
- [x] Terraform Lock (condition/target) #1554 
- [x] Terraform Provider (condition/target) #1554 
- [ ] Terraform Registry (source)
- [ ] Terraform Auto discovery

**Comment by olblak** [comment]
Yes it sounds good to me, I would consider the autodiscovery as a big bonus at the moment
